### ISSUE TITLE
Adding pre/post transaction triggers

### ISSUE THREAD

**Issue by n3npq** [OpenMandrivaAssociation/rpm#3]
OMA rpm-5.4.15 carries a patch to change the RPM state machine by adding pre/post transaction triggers. This issue splits out the discussion from issue #1.

Presumably (I don't know because I lack examples) the changes are being used to automate side effects (like running ldconfig for every library update/removal) "functionally" (i.e. based on library paths) to minimize execution overhead, presumably by using a glob/*RE on paths to fire a post transaction trigger. (Note this is just my guess without any explicit example).

This issue has been split off from other upgrade issues to discuss:

0) I need examples of intended usage to understand WTF?!?

1) If ldconfig as guessed above, there are alternative implementations, including embedding
ldconfig side effects directly in RPM, either through a libc API (as discussed with dropper in 2007), or
by swiping the glibc ldconfig implementation from glibc and maintaining in parallel directly in rpm.

2) if more than ldconfig (as guessed above) is being attempted in pre transaction triggers, like, say, running embedded scripts to pre-condition empty chroot/VM setup, then there are additional
implementations that likely need to be attempted, including
    a) ensuring that embedded interpreter du jour has executable-free binding support for common
        file system operations, including mkdev and/or mount, to support the scripting.
    b) a richer and more precise trigger namespace than glob/*RE on file paths will be needed

Otherwise, its very difficult (for me) to understand how pre transaction triggers can be used: there are huge problems running a "script" in an empty chroot, there are no helpers whatsoever, nor is it obvious how dependencies can be used to pre-pre-pre-populate the chroot with needed executables/modules generally.

**Comment by proyvind** [comment]
In *cooker, 'rpm -qa --triggers'...

There's a lot more triggers than ldconfig, requiring file arguments etc.
And these does a great job at solving scriptlet ordering issues as well! :D

For how I implemented it in rpm, I did a mistake however, with the size of argv[]*only being 4096, it would limit the number of file arguments passed to the trigger to 4096 (and most likely go out of bounds if exceeding), the arguments oughta rather be passed through stdin as done in pixel's original file triggers patch  and also in rpm.org.

Captain hindsight strikes again!

**Comment by n3npq** [comment]
pixel's original file trigger implementation has other flaws than what this issue is about (i.e. sortingbwhat patches are needed for rpm-5.4.15 -> rpm-5.4.17+ upgrade, specifically changes
to the transaction state machine by adding pre/post transaction triggers)

Open another issue if you wish to discuss pixel's (or rpm5.org) file triggers.

**Comment by tpgxyz** [comment]
@n3npq IIRC only systemd uses heavily triggers

https://github.com/OpenMandrivaAssociation/systemd/blob/master/systemd.spec#L559
https://github.com/OpenMandrivaAssociation/systemd/blob/master/systemd.spec#L792
https://github.com/OpenMandrivaAssociation/systemd/blob/master/systemd.spec#L802
https://github.com/OpenMandrivaAssociation/systemd/blob/master/systemd.spec#L838
https://github.com/OpenMandrivaAssociation/systemd/blob/master/systemd.spec#L850
More can be find in systemd.spec

These triggers really do help with packaging. I'd suggest to keep this functionality

**Comment by n3npq** [comment]
Thanks for examples.

Studying ... no matter what -- as proxy RPM maintainer -- you can do whatever you wish.

What I see (after 5 mins) however is an attempt to do automated configuration/cache management
using RPM.

Even if that "works", the abstractions are quite complicated, and a different mechanism than just RPM installation/erasure would improve the reliability of the attempted configuration management. JMHO, YMMV.

The fundamental flaws I see have to do with having only patterns on paths to trigger side-effects: a richer name space is needed to hide gory details (assuming you want stateless self-correcting functional package management, which is certainly a useful goal).

There is the subsidiary issue of whether globs or *RE's are used to trigger: RPM (as implemented upstream) uses globs. Offhand, I do not see any pattern that MUSTHAVE *RE's: Globs with curly brackets could easily represent, say,
`%triggerpostun -- ^%{_unitdir}/.*\.(service|socket|path|timer)$`
Adding support for BOTH *RE's and globs (at least) requires RPM documentation changes and tests so that the implementation in RPM is "supported" and "maintainable". And there is the further issue of *RE dialects ...

**Comment by tpgxyz** [comment]
@n3npq IMHO for me you can drop these triggers only if you help with replacing them with LUA scripts 👍

**Comment by n3npq** [comment]
The best solution would be to setup a daemon that watches the directories of interest for changes and Does The Right Thing outside of package management. JMHO.

Meanwhile I'm just your packaging bitch atm, pre/post transaction triggers in RPM if you must, I simply do not care how you choose to patch RPM.

**Comment by n3npq** [comment]
Q: Are there any other examples (other than systems) in OMA that use pre/post transaction triggers?

**Comment by tpgxyz** [comment]
IIRC only systemd uses triggers for critical stuff.

**Comment by n3npq** [comment]
OK, so pre/post transaction triggers are largely to automate systemd dynamic state changes by attaching/appending/running code snippets through triggers rather than (more traditionally) running a command in daemon package scriptlets.

Is that a fair statement of what is being attempted?

Q: Why not use fanotify/inotify instead of the RPM state machine? Both fanotify/inotify are more efficient than applying patterns to rpmdb keys to load/execute some script, and would seem to be more general to handling all changes to state, not just state changes during an install.

**Comment by proyvind** [comment]
The pre/post transaction triggers were implemented to provide same functionality as the corresponding pre/post ransaction scriptlets.

It's like 4-5 years ago I implemented these, with my activity and dealing with rpm5 & omv-cooker being minimal or the past few years, I'll have to try best remember this one from memory (the triggers were announced on explained on cooker list, but only files archives of are found at http://archive.openmandriva.org, I'll try see if I can get around to get archived cooker up with gmane gone and all...)..

IIRC the triggers served two purposes: guarantee that trggers are only run at end of transaction in order to guard against any ordering issues (uncertain about this one, can't come to think of any related to trigers ATM) and to avoid triggers from firing from each package with files matching, but rather just once with all matches from all packages at the end of transaction.

For pretransaction trigger, it was only added for completeness in order to complement the pre transaction scriptlet.


At least I think it was something like that..

Not involved in either projects, I'm comfortable 'nuff leaving it with that. :p

**Comment by n3npq** [comment]
After 5y, with the only reported usage case being systemd, there appears to be little general need for pre/post transaction triggers.

(aside)
Yes, there is a need to "Do no harm." to existing systemd packages while upgrading OMA RPM. Ensuring that systemd "works" is an entirely different packaging related issue than  changing the RPM state machine.

Go find consensus on the need and usage case for pre/post transaction triggers from the RPM Software Management Eco System if you wish a change to the RPM state machine.

Meanwhile I believe that identifying a different trigger implementation for systemd using fanotify/inotify 
and planning removal of pre/post transaction triggers in OMA RPM is the best medium term approach.

**Comment by n3npq** [comment]
List of pre/post transaction triggers from "rpm -qa --triggers"
[oma-triggerlist.gz](https://github.com/OpenMandrivaAssociation/rpm/files/778235/oma-triggerlist.gz)
